keps/sig-cloud-provider/2699-add-webhook-hosting-to-ccm/README.md

```diff
@@ -0,0 +1,541 @@
+disable or alter the functionality. This potentially makes it harder for a cloud 
+provider to enforce that this logic is being applied. It also means that there 
+needs to be a deployment mechanism for the webhook. It is left to the Cloud 
+Provider to determine if the need for an inline request is sufficient to override
+these concerns. The Cloud Provider can alternately use the controller route which 
+is not inline or use an actual admission controller, built into the APIServer.
+
+We are actually changing the framework which generates the CCM and not the CCM
+itself. It has been pointed out that it is not the role of the controller
+manager to run webhooks. Controller managers should run controllers and webhooks
+are not controllers. As we are modifying the framework, we should consider this
+as we can create two processes. The CCM which just has controllers in it. We
+can also create a Cloud Webhook Manager. That is being left as homework for the 
+Cloud Provider. However the sample CCM which demonstrates how this will be done
+will have both in the same sample to make it easy.
+
+It is wroth being aware that the CCM derives from the KCM. The KCM (and the CCM)
+predate efforts like controller runtime. The controller runtime is a good 
+reference as it is demonstrates that operators and webhooks can be successfully
+run inside the same binary. It further demonstrates that this is a pattern 
+which is understood and followed by a significant portion of the Kubernetes
+community. Having said that, we consider it more important to unify the KCM
+and CCM code bases, then to build on top controller runtime. We are not saying
+not to use anything from controller runtime. We are saying that if need to choose
+between unifying the KCM & CCM code and building with controller runtime, we will
+choose unifying the KCM & CCM code base.
+
+## Design Details
+
+A sample  of how the Builder pattern might look is:
+
+```
+cmOptions, err := options.NewCloudManagerOptions()
+if err != nil {
+klog.Fatalf("unable to initialize command options: %v", err)
+}
+fss := cliflag.NamedFlagSets{}
+cloudManagerBuilder := app.NewCloudManagerBuilder("name")
+cloudManagerBuilder.setOptions(cmOptions)
+cloudManagerBuilder.setFlags(fss)
+cloudManagerBuilder.registerWebhook(gvkList, handler)
+cloudManagerBuilder.registerWebhook(gvkSecondList, secondHandler)
+manager, err := cloudManagerBuilder(wait.NeverStop)
+if err != nil {
+klog.Fatalf("unable to construct cloud manager: %v", err)
+}
+err := command.start()
+```
+
+This will not alter the existing extension hooks in the controller manager 
+framework, as they are critical for backward compatibility. The builders
+are meant to be an abstraction layer on top to make the extensions easier to 
+use. So for the existing controller manager code you might see changes like:
+
+```
+cloudControllerManagerBuilder.registerController("nodeipamcontroller", handler)
```

Comment:
Is `handler` a `InitFunc` or an already-initiated `Controller` instance? If it is the latter, I believe the initialization process will need a foundational change.

Reply:
At this point it would almost certainly be a ControllerInitFuncConstructor. Having said that, this is meant to give a feeling for the builder pattern. It seems likely this will need to be adjusted some when we actually implement. We will of course then take those discoveries and feed them back into the KEP. This is after all still alpha.